dipy/reconst/mcsd.py

```diff
@@ -0,0 +1,347 @@
+import numpy as np
+import numpy.linalg as la
+from dipy.core import geometry as geo
+from dipy.data import default_sphere
+from dipy.reconst import shm
+from dipy.reconst.multi_voxel import multi_voxel_fit
+
+from dipy.utils.optpkg import optional_package
+cvx, have_cvxpy, _ = optional_package("cvxpy")
+
+SH_CONST = .5 / np.sqrt(np.pi)
+
+
+def multi_tissue_basis(gtab, sh_order, iso_comp):
+    """
+    Builds a basis for multi-shell CSD model.
+    """
+    if iso_comp < 1:
+        msg = ("Multi-tissue CSD requires at least 2 tissue compartments")
+        raise ValueError(msg)
+    r, theta, phi = geo.cart2sphere(*gtab.gradients.T)
+    m, n = shm.sph_harm_ind_list(sh_order)
+    B = shm.real_sph_harm(m, n, theta[:, None], phi[:, None])
+    B[np.ix_(gtab.b0s_mask, n > 0)] = 0.
+
+    iso = np.empty([B.shape[0], iso_comp])
+    iso[:] = SH_CONST
+
+    B = np.concatenate([iso, B], axis=1)
+    return B, m, n
+
+
+class MultiShellResponse(object):
+
+    def __init__(self, response, sh_order, shells):
+        """ Estimate Multi Shell response function for multiple tissues and
+        multiple shells.
+
+        Parameters
+        ----------
+        response : tuple or AxSymShResponse object
+            A tuple with two elements. The first is the eigen-values as an (3,)
+            ndarray and the second is the signal value for the response
+            function without diffusion weighting.  This is to be able to
+            generate a single fiber synthetic signal.
+        sh_order : ndarray
+        shells : int
+            Number of shells in the data
+        """
+        self.response = response
+        self.sh_order = sh_order
+        self.n = np.arange(0, sh_order + 1, 2)
+        self.m = np.zeros_like(self.n)
+        self.shells = shells
+        if self.iso < 1:
+            raise ValueError("sh_order and shape of response do not agree")
+
+    @property
+    def iso(self):
+        return self.response.shape[1] - (self.sh_order // 2) - 1
+
+
+def closest(haystack, needle):
+    diff = abs(haystack[:, None] - needle)
+    return diff.argmin(axis=0)
+
+
+def _inflate_response(response, gtab, n, delta):
+    if any((n % 2) != 0) or (n.max() // 2) >= response.sh_order:
+        raise ValueError("Response and n do not match")
+
+    iso = response.iso
+    n_idx = np.empty(len(n) + iso, dtype=int)
+    n_idx[:iso] = np.arange(0, iso)
+    n_idx[iso:] = n // 2 + iso
+
+    b_idx = closest(response.shells, gtab.bvals)
+    kernal = response.response / delta
+
+    return kernal[np.ix_(b_idx, n_idx)]
+
+
+def _basic_delta(iso, m, n, theta, phi):
+    """Simple delta function
+    Parameters
+    ----------
+    iso: int (optional)
```

Comment:
Says it's optional, but it doesn't seem to be optional in the function signature.

Reply:
That said, it would make sense to add error handling to make sure that it's at least 2.